File: libhwio/src/Interrupt.cpp
#include <stdint.h>
#include <ece643/libhwio/MMap.hpp>
#include <ece643/libhwio/Interrupt.hpp>

using namespace ece643::libhwio;

Interrupt::Interrupt(MMap &mmap) noexcept : mask(0), mmap(&mmap) {
}

void Interrupt::enable(uint32_t mask) {
    (*mmap)[0x200004] = this->mask |= mask;
}

void Interrupt::disable(uint32_t mask) {
    (*mmap)[0x200004] = this->mask &= ~mask;
    poll(~this->mask);
}

uint32_t Interrupt::poll(uint32_t mask) noexcept {
    uint32_t irq = (*mmap)[0x200000] & mask;
    (*mmap)[0x200008] = irq;
    return irq;
}

File: sim/src/main/kotlin/ece643/libhwio/Interrupt.kt
package ece643.libhwio

import ece643.sim.Button
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import javax.swing.JFrame

class Interrupt(window: JFrame, buttons: Array<Button>) {
    private var enableMask = 0
    private var activeMask = 0

    init {
        window.addWindowListener(object : WindowAdapter() {
            override fun windowClosing(e: WindowEvent?) {
                activeMask = activeMask.or(enableMask.and(0x000F0000))
            }
        })
        buttons.forEachIndexed { i, button ->
            val mask = 1.shl(if (i < 10) 9 - i else 29 - i)
            button.addActionListener {
                activeMask = activeMask.or(enableMask.and(mask))
            }
        }
    }

    fun enable(mask: Int) {
        try {
            enableMask = enableMask.or(mask)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun disable(mask: Int) {
        try {
            enableMask = enableMask.and(mask.inv())
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun poll(mask: Int): Int {
        try {
            val match = activeMask.and(mask)
            activeMask = activeMask.and(match.inv())
            return match
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return 0
    }
}
